src/animationLoop.js

```diff
@@ -0,0 +1,140 @@
+// This function is 3 to 5 times faster then Array.prototype.filter
+// but iterates in reverse and mutates the array. Worth it?
+function reverseFilter(array, callback, argument) {
+  let index = array.length;
+  while (index--) {
+    if (!callback(array[index], argument)) {
+      array.splice(index, 1);
+    }
+  }
+  return array;
+}
+
+function renderSubscriber(subscriber, alpha) {
+  subscriber.render(alpha, subscriber.value, subscriber.prevValue);
+  return subscriber.active;
+}
+
+const prototype = {
+  running: false,
+  shouldStop: false,
+  lastTime: 0,
+  accumulatedTime: 0,
+
+  setOptions: function setOptions(options) {
+    const animationLoop = this;
+
+    // Seconds
+    const timeStep = options.timeStep;
+
+    // Milliseconds
+    animationLoop.timeStep = timeStep * 1000;
+
+    animationLoop.timeScale = options.timeScale;
+    animationLoop.maxSteps = options.maxSteps;
+    animationLoop.getTime = options.getTime;
+    animationLoop.ticker = options.ticker;
+
+    animationLoop.step = function step(subscriber) {
+      if (subscriber.active) {
+        const value = subscriber.value;
+
+        subscriber.prevValue = value;
+        subscriber.value = subscriber.step(value, timeStep);
+      }
+    };
+
+    return animationLoop;
+  },
+
+  subscribe: function subscribe(value, step, render) {
+    const animationLoop = this;
+    const subscriber = {
+      value: value,
+      prevValue: value,
+      step: step,
+      render: render,
+      active: true,
+    };
+
+    animationLoop.state.push(subscriber);
+
+    return function unsubscribe() {
+      subscriber.active = false;
+    };
+  },
+
+  loop: function loop() {
+    const animationLoop = this;
+    const currentTime = animationLoop.getTime();
+
+    if (animationLoop.shouldStop) {
+      animationLoop.running = animationLoop.shouldStop = false;
+      return;
+    }
+
+    const timeStep = animationLoop.timeStep;
+    const frameTime = currentTime - animationLoop.lastTime;
+
+    animationLoop.lastTime = currentTime;
+    animationLoop.accumulatedTime += frameTime * animationLoop.timeScale;
+
+    if (animationLoop.accumulatedTime > timeStep * animationLoop.maxSteps) {
+      animationLoop.accumulatedTime = 0;
+    }
+
+    while (animationLoop.accumulatedTime > 0) {
+      animationLoop.state.forEach(animationLoop.step);
+      animationLoop.accumulatedTime -= timeStep;
+    }
+
+    // Render and filter in one iteration.
+    reverseFilter(
```

Comment:
What is the perf diff of the reverseFilter optimization vs. inlining renderSubscriber?

```js
const alpha = 1 + animationLoop.accumulatedTime / timeStep;
animationLoop.state = animationLoop.state.filter(subscriber => {
  subscriber.render(alpha, subscriber.value, subscriber.prevValue);
  return subscriber.active;
});
```